# 01. Importing Libraries

In [3]:
#importing libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [4]:
#importing products data from the project folder
df_prods = pd.read_csv (r'C:\Users\dmlos\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [2]:
#importing orders data from the project folder
df_ords = pd.read_csv (r'C:\Users\dmlos\Instacart Basket Analysis\02 Data\Original Data\orders.csv', index_col = False)

NameError: name 'pd' is not defined

## A.) Data Consistency Checks- Products Data

In [6]:
#investigating our data set
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.710542e+06  1.715486e+01   
std    9.875817e+05  9.875817e+05  9.875817e+05  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  8.552715e+05  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.710542e+06  1.100000e+01   
75%    2.565812e+06  2.565812e+06  2.565812e+06  2.300000e+01   
max    3.421082e+06  3.421083e+06  3.421083e+06  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [11]:
#Testing a Data Check on A Mixed Data Column

In [8]:
# Create a dataframe

df_test = pd.DataFrame()

In [9]:
# Create a mixed type column

df_test ['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# Check for mixed types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)


mix


In [16]:
# Fixing mixed type columns

df_test['mix'] = df_test['mix'].astype('str')

In [17]:
# Finding missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [19]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [20]:
df_prods.shape

(49693, 5)

In [21]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

In [28]:
#Cleaned Producsts Data Frame without Duplicates
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [32]:
path = (r'C:\Users\dmlos\Instacart Basket Analysis')

In [5]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'products_checked.csv'))

NameError: name 'df_prods_clean_no_dups' is not defined

## B.) Data Consistency Checks- Orders Data

In [34]:
# investigating our dataset

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.710542e+06  1.715486e+01   
std    9.875817e+05  9.875817e+05  9.875817e+05  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  8.552715e+05  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.710542e+06  1.100000e+01   
75%    2.565812e+06  2.565812e+06  2.565812e+06  2.300000e+01   
max    3.421082e+06  3.421083e+06  3.421083e+06  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

##### Question 2.) A preliminary run through of the numbers actually reveals a pretty logically-consistent set of data- the counts for each column are all the consistent, a good sign. The minimum, maximum, and quartile values make sense for the fields to which they apply (i..e orders_day_of_the_week, Order_hour_of_day, and days_since_prior_order.)

In [36]:
# Check for mixed types

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

In [37]:
# Check for missing values

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### Question 5.) There are 206,209 missing values for days_since_prior_order. I believe this result comes from the fact that there are quite a few entries for same-day orders (i.e. 0 days since prior order). Because those orders were made the same day, they were left blank in the "days since prior order" field.

##### Question 6.) For the blank values, I would impute a value of '0'.

In [39]:
#Impute values of '0' for missing or blank values in the 'days_since_prior_order' column.

df_ords['days_since_prior_order'].fillna (0, inplace = True)

In [41]:
#Look at df_ords with imputed values.

df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329  2539329    prior             1   
1                 1   2398795  2398795    prior             2   
2                 2    473747   473747    prior             3   
3                 3   2254736  2254736    prior             4   
4                 4    431534   431534    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710  2266710    prior            10   
3421079     3421079   1854736  1854736    prior            11   
3421080     3421080    626363   626363    prior            12   
3421081     3421081   2977660  2977660    prior            13   
3421082     3421082    272231   272231    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     0.0  
1                         3                  7                    15.0  
2                         3                 12                    21.0  
3                         4                  7                    29.0  
4                         4                 15                    28.0  
...                     ...                ...                     ...  
3421078                   5                 18                    29.0  
3421079                   4                 10                    30.0  
3421080                   1                 12                    18.0  
3421081                   1                 12                     7.0  
3421082                   6                 14                    30.0  

[3421083 rows x 8 columns]

##### Question 7.)  In the ".duplicated()" function below, no duplicate rows were found.

In [42]:
#Looking for duplicate rows in the df_ords dataset.

df_dups = df_ords[df_ords.duplicated()]

In [43]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [44]:
df_ords.shape

(3421083, 8)

##### Question 8.) Truthfully, I didn't find any "full duplicate" values, so, there was no action needed to address duplicates.

In [46]:
path = (r'C:\Users\dmlos\Instacart Basket Analysis')

In [47]:
df_ords.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_checked.csv'))

In [4]:
df_prods.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'products_checked(3).csv'))

NameError: name 'df_prods' is not defined